pkg/binder/binder_unit.go

```diff
@@ -267,79 +267,46 @@ func (unitInfo *bindingUnitInfo) GetNewTasksOnNode(nodeName string) []*runningUn
 	return result
 }
 
-func (unitInfo *bindingUnitInfo) GetReadyAndWaitingTasks() []*checkResult {
+func (unitInfo *bindingUnitInfo) getTasks(tasks map[types.UID]*checkResult) []*checkResult {
 	unitInfo.mu.Lock()
 	defer unitInfo.mu.Unlock()
 
-	len := len(unitInfo.readyTasks) + len(unitInfo.waitingTasks)
-	if len == 0 {
+	length := len(tasks)
+	if length == 0 {
 		return nil
 	}
 
-	result := make([]*checkResult, len)
+	result := make([]*checkResult, length)
 	i := 0
-	for _, cr := range unitInfo.readyTasks {
-		result[i] = cr
-		i++
-	}
-	for _, cr := range unitInfo.waitingTasks {
+	for _, cr := range tasks {
 		result[i] = cr
 		i++
 	}
 	return result
 }
 
-func (unitInfo *bindingUnitInfo) GetReadyTasks() []*checkResult {
-	unitInfo.mu.Lock()
-	defer unitInfo.mu.Unlock()
-
-	len := len(unitInfo.readyTasks)
-	if len == 0 {
-		return nil
+func (unitInfo *bindingUnitInfo) GetReadyAndWaitingTasks() []*checkResult {
```

Comment:
We should hold RLock there to avoid concurrent reading and writing, rather just in `getTasks`